### PR TITLE
[Common] Add error message for tensor common functions

### PR DESCRIPTION
**Pull request by gichan-jang**
Add error message for tensor common functions.

Signed-off-by: gichan <gichan2.jang@samsung.com>

**Self evaluation:**
1. Build test: [ ]Passed [ ]Failed [*]Skipped
2. Run test: [ ]Passed [ ]Failed [*]Skipped

### CODE REVIEW

File: gst/nnstreamer/tensor_common.c [hunk 1]
```diff
@@ -1538,24 +1575,43 @@ gst_tensor_meta_info_validate (GstTensorMetaInfo * meta)
...
+    nns_logw ("Failed to validate tensor meta info. invalid media type: %d.",
+        meta->media_type);
+    goto error;
+  }
 
   return TRUE;
+
+error:
+  nns_logw
+      ("NNStreamer internal error. Plese post issue on https://github.com/nnstreamer/nnstreamer/issues.");
```
Comment (dongju-chae) [inline gst/nnstreamer/tensor_common.c:1613]: Typo detected :)

**Review by dongju-chae — APPROVED** [review]
LGTM

File: gst/nnstreamer/tensor_common.c (hunk 1 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_common.c]: Updated. Thanks!

File: gst/nnstreamer/tensor_common.c [hunk 2]
```diff
@@ -485,7 +499,13 @@ gst_tensors_info_is_equal (const GstTensorsInfo * i1, const GstTensorsInfo * i2)
...
   g_return_val_if_fail (i2 != NULL, FALSE);
 
-  if (i1->num_tensors != i2->num_tensors || i1->num_tensors == 0) {
+  if (!gst_tensors_info_validate (i1) || !gst_tensors_info_validate (i2)) {
+    return FALSE;
+  }
+
+  if (i1->num_tensors != i2->num_tensors) {
+    nns_logw ("Tensors info is not equal. the number of tensors: %d vs %d. ",
+        i1->num_tensors, i2->num_tensors);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_common.c:508]: This is info/debug, not warning.

File: gst/nnstreamer/tensor_common.c [hunk 3]
```diff
@@ -295,11 +298,20 @@ gst_tensor_info_is_equal (const GstTensorInfo * i1, const GstTensorInfo * i2)
...
+        GST_STR_NULL (gst_tensor_get_type_string (i2->type)));
     return FALSE;
   }
 
   for (i = 0; i < NNS_TENSOR_RANK_LIMIT; i++) {
     if (i1->dimension[i] != i2->dimension[i]) {
+      gchar *dim_str1 = gst_tensor_get_dimension_string (i1->dimension);
+      gchar *dim_str2 = gst_tensor_get_dimension_string (i2->dimension);
+      nns_logw ("Tensor info is not equal. Given tensor dimensions %s vs %s",
+          dim_str1, dim_str2);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_common.c:312]: This should be debug, not warning.

File: gst/nnstreamer/tensor_common.c [hunk 4]
```diff
@@ -274,6 +274,9 @@ gst_tensor_info_validate (const GstTensorInfo * info)
   g_return_val_if_fail (info != NULL, FALSE);
 
   if (info->type == _NNS_END) {
+    nns_logw ("Failed to validate tensor info. type: %s. "
+        "Please specify tensor type. e.g., type=uint8 ",
+        GST_STR_NULL (gst_tensor_get_type_string (info->type)));
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_common.c:279]: This should be debug, not warning. You don't know the caller's context.

File: gst/nnstreamer/tensor_common.c [hunk 5]
```diff
@@ -794,11 +814,18 @@ gst_tensors_config_validate (const GstTensorsConfig * config)
 
   /* framerate (numerator >= 0 and denominator > 0) */
   if (config->rate_n < 0 || config->rate_d <= 0) {
+    nns_logw ("Failed to validate tensors config. framerate: %d/%d. "
+        "framerate should be numerator >= 0 and denominator > 0.",
+        config->rate_n, config->rate_d);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_common.c:819]: This should be debug, not warning. You don't know the caller's context.

The caller wants to know if it's valid or not. The caller might intentionally test invalid tensors (to do something else if it's invalid).

File: gst/nnstreamer/tensor_common.c [hunk 6]
```diff
@@ -1538,24 +1575,43 @@ gst_tensor_meta_info_validate (GstTensorMetaInfo * meta)
...
+        meta->media_type);
+    goto error;
+  }
 
   return TRUE;
+
+error:
+  nns_logw
+      ("NNStreamer internal error. Please post issue on https://github.com/nnstreamer/nnstreamer/issues.");
+  return FALSE;
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_common.c:1614]: Be careful. You don't know if returning FALSE is an issue or not. The caller knows if FALSE is an issue or not.

File: gst/nnstreamer/tensor_common.c [hunk 7]
```diff
@@ -1212,6 +1244,11 @@ gst_tensor_dimension_is_valid (const tensor_dim dim)
 
   for (i = 0; i < NNS_TENSOR_RANK_LIMIT; ++i) {
     if (dim[i] == 0) {
+      gchar *dim_str = gst_tensor_get_dimension_string (dim);
+      nns_logw ("Failed to validate tensor dimension. Given dimension: %s"
+          "The dimension string should be in the form of d1:d2:d3:d4, "
+          "d1:d2:d3, d1:d2, or d1. Here, dN is a positive integer.", dim_str);
+      g_free (dim_str);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_common.c:1251]: reconsider the log level.

**Comment by myungjoo** [comment]
ERROR: it's a problem critical enough to stop the pipeline
WARNING: it's a problem. not we may keep running as long as we notify the programmer.

in a function, ```is_valid(a)```, ```a``` being invalid is NOT a problem. The caller may think it's a problem, but that's what the caller needs to decide.

**Comment by gichan-jang** [comment]
> ERROR: it's a problem critical enough to stop the pipeline WARNING: it's a problem. not we may keep running as long as we notify the programmer.
> 
> in a function, `is_valid(a)`, `a` being invalid is NOT a problem. The caller may think it's a problem, but that's what the caller needs to decide.

Thanks. I changed the log level to debug.

**Review by jaeyun-jung — APPROVED** [review]

**Review by myungjoo — APPROVED** [review]

**Pull request merged**